### PR TITLE
mqtt protocol binding

### PR DESCRIPTION
**Pull request by yanmxa**
Initial implementation of MQTT binding, using https://github.com/eclipse/paho.golang which supports MQTT V5

Note: This is just a basic implementation; some new features which are brought by MQTT 5.0 and other enhancements will be done in future PRs.

resolved: https://github.com/cloudevents/sdk-go/issues/901

### CODE REVIEW

**Comment by duglin** [comment]
@yanmxa thanks.  Quick scan looks nice - will review more...

ping @lionelvillard and @embano1 for their review too.

File: samples/mqtt/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+MQTT samples
+
+To run the samples, you need a running MQTT broker.
+
+To run a sample MQTT broker using docker:
+
+```bash
+echo "listener 1883 0.0.0.0
+allow_anonymous true" > samples/mosquitto.conf
```
Comment (embano1) [inline samples/mqtt/README.md:9]: nit: remove echo newline here

File: samples/mqtt/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+docker run --rm --name mosquitto -p 1883:1883 -v "$(pwd)/samples/mosquitto.conf:/mosquitto/config/mosquitto.conf" eclipse-mosquitto
```
Comment (embano1) [inline samples/mqtt/README.md:10]: looks like `mosquitto.conf` is not part of the samples here? where is this command supposed to be run (since it refs `pwd`)?

File: samples/mqtt/sender/main.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	if err != nil {
+		log.Fatalf("failed to create client, %v", err)
+	}
+
+	for i := 0; i < count; i++ {
+		e := cloudevents.NewEvent()
+		e.SetID(uuid.New().String())
+		e.SetType("com.cloudevents.sample.sent")
+		e.SetSource("https://github.com/cloudevents/sdk-go/samples/mqtt/sender")
+		_ = e.SetData(cloudevents.ApplicationJSON, map[string]interface{}{
```
Comment (embano1) [inline samples/mqtt/sender/main.go:50]: nit: you check `err` always, just not here?

File: samples/mqtt/sender/main.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	clientConfig := &paho.ClientConfig{
+		ClientID: "sender-client-id",
+		Conn:     conn,
+	}
+	cp := &paho.Connect{
+		KeepAlive:  30,
+		CleanStart: true,
+	}
+
+	p, err := cemqtt.New(context.TODO(), clientConfig, cp, "test-topic", nil, 0, false)
```
Comment (embano1) [inline samples/mqtt/sender/main.go:34]: nit: create a `ctx` and reuse instead of TODO

File: samples/mqtt/receiver/main.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		log.Fatalf("failed to create protocol: %s", err.Error())
+	}
+	defer p.Close(context.Background())
+
+	c, err := cloudevents.NewClient(p)
+	if err != nil {
+		log.Fatalf("failed to create client, %v", err)
+	}
+
+	log.Printf("will listen consuming topic test-topic\n")
```
Comment (embano1) [inline samples/mqtt/receiver/main.go:39]: nit: wording?

File: samples/mqtt/receiver/main.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	clientConfig := &paho.ClientConfig{
+		ClientID: "receiver-client-id",
+		Conn:     conn,
+	}
+	cp := &paho.Connect{
+		KeepAlive:  30,
+		CleanStart: true,
+	}
+
+	p, err := cemqtt.New(context.TODO(), clientConfig, cp, "test-topic", []string{"test-topic"}, 0, false)
```
Comment (embano1) [inline samples/mqtt/receiver/main.go:28]: question: is sender topic here required, or would `""` be ok?

Comment (yanmxa) [reply samples/mqtt/receiver/main.go:28]: The senderTopic isn't necessary for the receiver, `""` is ok.

File: samples/mqtt/README.md (hunk 1 above)
Comment (yanmxa) [reply samples/mqtt/README.md]: Done!

File: samples/mqtt/README.md (hunk 2 above)
Comment (yanmxa) [reply samples/mqtt/README.md]: @embano1 Sorry for the confusion. I just updated the doc with a simpler command, so that it can run any path.

File: samples/mqtt/sender/main.go (hunk 4 above)
Comment (yanmxa) [reply samples/mqtt/sender/main.go]: Done!

File: samples/mqtt/receiver/main.go (hunk 5 above)
Comment (yanmxa) [reply samples/mqtt/receiver/main.go]: Done!

File: protocol/mqtt_paho/v2/message_test.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+			}
+		})
+	}
+}
+
+func TestReadBinary(t *testing.T) {
+	// msg := &paho.Publish{
+	// 	Payload:    []byte("{hello:world}"),
+	// 	Properties: &paho.PublishProperties{ContentType: event.ApplicationCloudEventsJSON},
+	// }
```
Comment (embano1) [inline protocol/mqtt_paho/v2/message_test.go:48]: remove?

File: protocol/mqtt_paho/v2/protocol.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
+package mqtt_paho
```
Comment (embano1) [inline protocol/mqtt_paho/v2/protocol.go:1]: This file has no tests, can we get some coverage here e.g., using integration tests?

File: protocol/mqtt_paho/v2/write_message.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	_, err := binding.Write(
+		ctx,
+		m,
+		structuredWriter,
+		binaryWriter,
+		transformers...,
+	)
+	return err
+}
```
Comment (embano1) [inline protocol/mqtt_paho/v2/write_message.go:29]: ideally we have tests for this one

**Comment by embano1** [comment]
Overall looking good @yanmxa We just need some more tests, especially around attribute/extension handling (verification) and writing messages.

File: protocol/mqtt_paho/v2/message.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+// Check if .Message implements binding.Message
+var (
+	_ binding.Message               = (*Message)(nil)
+	_ binding.MessageMetadataReader = (*Message)(nil)
+)
+
+func NewMessage(msg *paho.Publish) *Message {
+	var f format.Format = nil
+	var v spec.Version = nil
```
[the remark this answers is not in the document]
Comment (yanmxa) [reply protocol/mqtt_paho/v2/message.go]: Done!

File: protocol/mqtt_paho/v2/message.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	if msg.Properties != nil {
+		// Use Content-type attr to determine if message is structured and
+		// set format.
+		if s := msg.Properties.ContentType; format.IsFormat(s) {
+			f = format.Lookup(s)
+		}
+		// Binary v0.3:
```
[the remark this answers is not in the document]
Comment (yanmxa) [reply protocol/mqtt_paho/v2/message.go]: This is a legacy comment without much meaning. I just deleted it from here~

File: protocol/mqtt_paho/v2/message_test.go (hunk 7 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/message_test.go]: Done.

File: protocol/mqtt_paho/v2/write_message.go (hunk 9 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/write_message.go:29]: As above.

File: protocol/mqtt_paho/v2/protocol.go (hunk 8 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/protocol.go:1]: I had planned to add an integration test in future PR, which might need to set some environment in the workflow.
If you think it's better to put it here, I will and these tests tomorrow~

File: protocol/mqtt_paho/v2/protocol.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+	topic := cecontext.TopicFrom(ctx)
+	if topic == "" {
+		topic = t.senderTopic
+	}
+
+	msg := &paho.Publish{
+		QoS:    t.qos,
+		Retain: t.retained,
+		Topic:  topic,
```
Comment (skeeey) [inline protocol/mqtt_paho/v2/protocol.go:74]: can we allow to add properties here?

File: protocol/mqtt_paho/v2/protocol.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+	_, err = t.client.Publish(ctx, msg)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (t *Protocol) OpenInbound(ctx context.Context) error {
+	t.openerMutex.Lock()
```
Comment (skeeey) [inline protocol/mqtt_paho/v2/protocol.go:89]: why need lock here?

File: protocol/mqtt_paho/v2/protocol.go (hunk 12 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/protocol.go:74]: We can add properties in the event extensions and it will be propagated to the `paho.Pulish.Properties`.

File: protocol/mqtt_paho/v2/protocol.go (hunk 13 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/protocol.go:89]: This method is to bootstrap a `Receiver/Responder`, it's a blocking call and ctx is used to stop the Inbound message Receiver/Responder. I think that can only be started once, that's the reason a lock is placed here. 

Do I understand correctly? @duglin  If not, please help me to clarify that. Thanks!

Comment (embano1) [reply protocol/mqtt_paho/v2/protocol.go:89]: Yes, other protocol implementations (besides AMQP IIUC) implement 

```go
// Opener is the common interface for things that need to be opened.
type Opener interface {
	// OpenInbound is a blocking call and ctx is used to stop the Inbound message Receiver/Responder.
	// Closing the context won't close the Receiver/Responder, aka it won't invoke Close(ctx).
	OpenInbound(ctx context.Context) error
}
```

in the same way using a mutex. It's a bit confusing as blocking can mean many things, in this project it seems to actually mimic a singleton.

File: protocol/mqtt_paho/v2/protocol.go (hunk 8 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/protocol.go:1]: @embano1 I added some integration tests today, PTAL~

**Comment by yanmxa** [comment]
> Overall looking good @yanmxa We just need some more tests, especially around attribute/extension handling (verification) and writing messages.

Sure! I added some unit tests and integration test for protocol and writing messages today. Thanks for your review!

File: protocol/mqtt_paho/v2/write_message.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+func (b *pubMessageWriter) SetAttribute(attribute spec.Attribute, value interface{}) error {
+	if attribute.Kind() == spec.DataContentType {
+		if value == nil {
+			b.removeProperty(contentType)
+		}
+		s, err := types.Format(value)
+		if err != nil {
+			return err
+		}
+		b.addProperty(contentType, s)
```
Comment (embano1) [inline protocol/mqtt_paho/v2/write_message.go:89]: unchecked `error` (IIUC, this is copy and paste from other implementations, but perhaps worth checking it here)

**Comment by embano1** [comment]
Looks good @yanmxa , just one final minor nit

File: protocol/mqtt_paho/v2/write_message.go (hunk 14 above)
Comment (yanmxa) [reply protocol/mqtt_paho/v2/write_message.go]: @embano1 Nice Catch! Done~

**Review by embano1 — APPROVED** [review]
GREAT JOB!

**Review by morvencao — COMMENTED** [review]
looks good, thanks @yanmxa

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+			})
+	})
+}
+
+func testSenderReceiver(t testing.TB, ctx context.Context) (bindings.Sender, bindings.Receiver, func()) {
+	topicName := "test-ce-client-" + uuid.New().String()
+
+	broker := os.Getenv("TEST_MQTT_BROKER")
+	if broker == "" {
+		broker = "127.0.0.1:1883"
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:67]: nit: make const

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func testSenderReceiver(t testing.TB, ctx context.Context) (bindings.Sender, bindings.Receiver, func()) {
+	topicName := "test-ce-client-" + uuid.New().String()
+
+	broker := os.Getenv("TEST_MQTT_BROKER")
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:65]: nit: make const

Comment (embano1) [reply test/integration/mqtt_paho_binding/mqtt_test.go:65]: also, where does this env come from?

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+	receiverConn, err := net.Dial("tcp", broker)
+	require.NoError(t, err)
+	r, err := mqtt_paho.New(ctx, &paho.ClientConfig{
+		Conn: receiverConn,
+	}, cp, topicName, []string{topicName}, 0, false)
+	require.NoError(t, err)
+	require.NotNil(t, r)
+
+	go func() {
+		require.NoError(t, r.OpenInbound(context.TODO()))
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:91]: failing in a separate gorouine is not allowed: https://pkg.go.dev/testing#T.FailNow

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+	}()
+
+	time.Sleep(5 * time.Millisecond) // let the receiver ready
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:94]: you can avoid this (anti-pattern) and the issue with failing in a goroutine above using channels and a ctx with timeout. please rewrite accordingly 🙏🏻

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+				eventOut := MustToEvent(t, ctx, out)
+				assert.Equal(t, binding.EncodingBinary, out.ReadEncoding())
+				eventOut = ConvertEventExtensionsToString(t, eventOut)
+
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func testSenderReceiver(t testing.TB, ctx context.Context) (bindings.Sender, bindings.Receiver, func()) {
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:62]: lower case test?

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 15 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Done~

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 16 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: I used the environment for local tests, and now remove it from here.

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 17 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Nice Catch! Remove the assert from the goroutine.

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 19 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Sorry for the unclear naming. It's only for getting the sender and receiver, so change it to `getSenderReceiver`.

**Comment by yanmxa** [comment]
@embano1 Thank you very much for checking the code so carefully, which allows me to avoid similar problems in the future.

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	. "github.com/cloudevents/sdk-go/v2/binding/test"
+	. "github.com/cloudevents/sdk-go/v2/test"
+)
+
+const (
+	Broker = "127.0.0.1:1883"
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:25]: nit: no need to export IMHO

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+		select {
+		case <-ctx.Done():
+			require.Fail(t, "timeout waiting for event")
+			return
+		case receiveMsg := <-receiveChan:
+			require.NoError(t, receiveMsg.err)
+			outAssert(receiveMsg.msg)
+			return
+		case <-time.After(2 * time.Second):
+			// the receiver mightn't be ready before the sender send the message, so we retry
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:91]: nit: this implementation wasn't obvious. consider adding moving the comment below and add a `continue` statement so it's clear you want to retry (loop)

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			require.Fail(t, "timeout waiting for event")
+			return
+		case receiveMsg := <-receiveChan:
+			require.NoError(t, receiveMsg.err)
+			outAssert(receiveMsg.msg)
+			return
+		case <-time.After(2 * time.Second):
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:90]: nit: not an issue in this test, but time.After can leak (consider using Ticker). check google for leaks ,)

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+			func(out binding.Message) {
+				assert.Equal(t, binding.EncodingStructured, out.ReadEncoding())
+				eventOut := MustToEvent(t, ctx, out)
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func TestSendBinaryMessageToBinary(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:53]: nit: defer cancel missing

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+	s, r, c, err := getSenderReceiver(ctx, cancel)
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:54]: nit: generally I find this code hard to read, not blocking you here though

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+	require.NoError(t, err)
+	defer c()
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:56]: especially this one is just ???

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+
+	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
+		eventIn = ConvertEventExtensionsToString(t, eventIn)
+		in := MustCreateMockStructuredMessage(t, eventIn)
+
+		sendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingStructured), s, r, in,
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:43]: do we really need that many levels of indirection? can't we keep the whole code within the test function? I have a feeling we're overcomplicating the test

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 21 above)
Comment (embano1) [reply test/integration/mqtt_paho_binding/mqtt_test.go:91]: also, since you retry anyways, do we need all these sleeps?

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+			return
+		case <-time.After(2 * time.Second):
+			// the receiver mightn't be ready before the sender send the message, so we retry
+			result := sender.Send(ctx, in)
+			require.NoError(t, result)
+		}
+	}
+}
+
+func getSenderReceiver(ctx context.Context, cancel context.CancelFunc) (
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:98]: that whole code is really hard to follow (and thus review)

Comment (embano1) [reply test/integration/mqtt_paho_binding/mqtt_test.go:98]: may be want to go back to the drawing board, thinking about what you're trying to accomplish and then simplify within one Test function?

**Comment by embano1** [comment]
> @embano1 Thank you very much for checking the code so carefully, which allows me to avoid similar problems in the future.

No worries, that's my duty here ;) You're doing great work! Just try to keep the code simple and clean - your future self (and other contributors) will appreciate it ;)

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 21 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go:91]: Yes, I think we need to wait here, otherwise, it will throw the error message like this:
```bash
{"level":"error","ts":1688780919.884921,"logger":"fallback","caller":"client/invoker.go:77","msg":"call to Invoker.Invoke(...) has panicked: send on closed channel",
```

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 22 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go:90]: Thank you for the reminder @embano1! 
It really has a race condition to cause the memory leak when using `time.After(2 * time.Second) `in such a scenario.

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 24 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go:54]: Okay, to make it clear, I modified it to avoid the shorthand。

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 26 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go:43]: I try to reuse `sendReceive` in `TestSendStructuredMessageToStructured` and `TestSendBinaryMessageToBinary`. 

Of course, we can expand these codes and put them here. In this case, there will be some redundant codes in these two tests.
If you prefer to expand it, please reply directly to me and I can remove this layer from here.

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 20 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Done.

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 22 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Done!

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 23 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Done!

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 25 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: Done!

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 27 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: As mentioned above, I have now changed it to try to avoid using this seemingly more complicated function.

**Comment by yanmxa** [comment]
@embano1 Thanks! It was indeed too complicated to write tests with reference to other tests before, but it should be simplified a lot now compared to before.

File: test/integration/mqtt_paho_binding/mqtt_test.go [hunk 28]
```diff
@@ -2,139 +2,130 @@ package mqtt_paho_binding
...
+						require.Fail(t, "timeout waiting for event")
+						return
+					case receiveMsg := <-receiveChan:
+						require.NoError(t, receiveMsg.err)
+						outMessage := receiveMsg.msg
+						assert.Equal(t, tt.expectedEncoding, outMessage.ReadEncoding())
+						outEvent := MustToEvent(t, ctx, outMessage)
+						AssertEventEquals(t, inEvent, ConvertEventExtensionsToString(t, outEvent))
+						return
+					default:
```
Comment (embano1) [inline test/integration/mqtt_paho_binding/mqtt_test.go:85]: nit: just be aware that this is a hot loop now, ideally should also be a ticker, right?

**Review by embano1 — APPROVED** [review]
Approving, just one minor nit if you feel it's worth the change

File: test/integration/mqtt_paho_binding/mqtt_test.go (hunk 28 above)
Comment (yanmxa) [reply test/integration/mqtt_paho_binding/mqtt_test.go]: @embano1 Yes, It's a hot loop! 
Since the `default` case causes a bit of confusion here, I turned it into an explicit ticker.

**Review by embano1 — APPROVED** [review]
LGTM

Feel free to squash and we should be good to go cc/ @duglin

**Comment by lionelvillard** [comment]
Overall it looks good, though I'm not an MQTT expert so there is so much I can review. Please add the copyright header to all files and then I think we can merge this PR.

**Comment by lionelvillard** [comment]
looks good. @duglin, ok with you?

**Comment by lionelvillard** [comment]
@embano1 approved it
LGTM
